experiment/secret-sync-controller/main_test.go

```diff
@@ -1,295 +1,435 @@
+									},
+								},
+							},
 						},
 					},
-					"ns2": map[string]map[string][]byte{
-						"secret2": map[string][]byte{
-							"key2": []byte("valueOfSecret2"),
+					{
+						Name: "ns-b",
+						Secrets: []client.K8sSecretSetup{
+							{
+								Name: "secret-b",
+							},
 						},
 					},
-				},
-				SecretManagerSecret: map[string]map[string][]byte{
-					"projA": map[string][]byte{
-						"secretA": []byte("valueOfSecretA"),
-					},
-					"projB": map[string][]byte{
-						"secretB": []byte("valueOfSecretB"),
+					{
+						Name: "ns-c",
 					},
 				},
+			},
+			expectErr: true,
+		},
+	}
+	for _, tt := range tests {
+		testname := fmt.Sprintf("%s", tt.config)
+		t.Run(testname, func(t *testing.T) {
+
+			// prepare clients
+			k8sClientset, err := client.NewK8sClientset()
+			if err != nil {
+				t.Errorf("New kubernetes client failed: %s", err)
 			}
-			err := Sync(client, tt.input)
-			if tt.expectErr {
-				if err == nil {
-					t.Errorf("Expected error but got none.")
-				}
-			} else {
-				if err != nil {
-					t.Errorf("Expected none error but got \"Error: %s\".", err)
-				}
-			}
+			secretManagerClient, err := client.NewSecretManagerClient(context.Background())
 			if err != nil {
-				return
+				t.Errorf("New Secret Manager client failed: %s", err)
 			}
-			secret, _ := client.GetKubernetesSecret(tt.input.Destination)
-			if !reflect.DeepEqual(secret.Data, tt.want) {
-				t.Errorf("Expected %s but got %s.", tt.want, secret.Data)
+			clientInterface := &client.Client{
+				K8sClientset:        *k8sClientset,
+				SecretManagerClient: *secretManagerClient,
 			}
-		})
-	}
-}
 
-/*
-func TestGetKubernetesSecret(t *testing.T) {
+			controller := &SecretSyncController{
+				Client:  clientInterface,
+				Config:  &tt.config,
+				RunOnce: true,
+			}
 
-	var tests = []struct {
-		input KubernetesSpec
-		want  []byte
-		error bool
-	}{
-		{
-			input: KubernetesSpec{
-				Namespace: "ns1",
-				Secret:    "secret1",
-				Key:       "key1",
-			},
-			want:  []byte("valueOfSecret1"),
-			error: false,
-		},
-		{
-			input: KubernetesSpec{
-				Namespace: "ns2",
-				Secret:    "secret2",
-				Key:       "key2",
-			},
-			want:  []byte("valueOfSecret2"),
-			error: false,
-		},
-		{
-			input: KubernetesSpec{
-				Namespace: "ns1",
-				Secret:    "secret2",
-				Key:       "key1",
-			},
-			want:  []byte{},
-			error: true,
-		},
-		{
-			input: KubernetesSpec{
-				Namespace: "ns1",
-				Secret:    "secret1",
-				Key:       "key0",
-			},
-			want:  []byte{},
-			error: true,
-		},
-	}
-	for _, tt := range tests {
-		testname := fmt.Sprintf("%s", tt.input)
-		t.Run(testname, func(t *testing.T) {
-			secret, err := mockClient.GetKubernetesSecret(tt.input)
-			if tt.error {
-				if err == nil {
-					t.Errorf("Expected error but got none.")
-				}
-			} else {
-				if err != nil {
-					t.Errorf("Expected none error but got \"Error: %s\".", err)
-				}
+			err = controller.Client.SetupTesting(setup)
```

Comment:
As discussed above, this should not be a method on Client, but instead perhaps a method defined in this file